wallet: Make scan / abort status private to CWallet

Empact opened the pull request:
By returning `ScanResult` from `RescanFromTime` and reporting the effectuality of `AbortScan`.

~And consolidate rpc-level error handling across `RescanFromTime` and
`ScanForWalletTransactions`.~

~Note this changes the `rescanblockchain` scan failure error from
`RPC_MISC_ERROR` to `RPC_WALLET_ERROR`, which seems more appropriate and matches the
behavior from the rpcdump methods.~

This follows up on #13076.

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -557,7 +553,7 @@ UniValue importwallet(const JSONRPCRequest& request)
     }
 
-    int64_t nTimeBegin = 0;
+    int64_t nTimeBegin = TIMESTAMP_MIN;
```
This is a dead store since `nTimeBegin` is unconditionally written to on L573.

Perhaps you could drop the dead store and instead move the `nTimeBegin = chainActive.Tip()->GetBlockTime();` to the line after `LOCK(pwallet->cs_wallet);`?

Empact replied:
`nTimeBegin` is used outside that block for the rescan, line 646.
I prefer to leave this as it seems harmless. Alternative is leaving it uninitialized.

practicalswift replied:
What I meant is that the current code is:

```
int64_t nTimeBegin = TIMESTAMP_MIN;
nTimeBegin = chainActive.Tip()->GetBlockTime();
RescanWallet(*pwallet, reserver, nTimeBegin, false /* update */);
```

Thus the first assignment is redundant since it is overwritten in all possible execution paths.

There are no execution paths where dropping the first assignment would result in an uninitialised variable being read.

Empact committed:
Return ScanResult from CWallet::RescanFromTime

Lookup the failed time just in the case it's needed.

Empact committed:
Make scan / abort status private to CWallet

Drop CWallet::IsAbortingRescan by reporting the abort
effectuality from CWallet::AbortScan.

Empact commented:
Rebased and substantially refocused this PR to:
* Return `ScanResult` from `RescanFromTime`
* Add `ScanResult::FailedTime` to report the best non-failing time - beneficial also because some callers are uninterested in the time
* Drop `IsScanning` and `IsAbortingRescan` from `CWallet`, as all scan/abort status is now returned as a result from specific operations.

Empact commented:
Rebased

jb55 commented:
looks like a nice refactor

utACK 4b4cf7cc28347b2d537be89a412093dbd62ea83e

Empact commented:
Rebased, and restored `IsScanning` as it's needed for d3e8458365

jb55 commented:
not sure why github tends to screw up range-diffs sometimes but here's the last one:

```
$ git range-diff 4b4cf7c~2..4b4cf7c 469117d~2..469117d
```
```diff
1:  99a72728a6 = 1:  c82f6c7806 Return ScanResult from CWallet::RescanFromTime
2:  4b4cf7cc28 ! 2:  469117d55c Make scan / abort status private to CWallet
@@ -2,7 +2,7 @@
 
     Make scan / abort status private to CWallet
 
-    Drop CWallet::IsAbortingRescan and IsScanning by reporting the abort
+    Drop CWallet::IsAbortingRescan by reporting the abort
     effectuality from CWallet::AbortScan.
 
  diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
@@ -54,8 +54,7 @@
       */
 -    void AbortRescan() { fAbortRescan = true; }
 -    bool IsAbortingRescan() { return fAbortRescan; }
--    bool IsScanning() { return fScanningWallet; }
 +    bool AbortRescan();
- 
-     /**
-      * keystore implementation
+     bool IsScanning() { return fScanningWallet; }
+     int64_t ScanningDuration() const { return fScanningWallet ? GetTimeMillis() - m_scanning_start : 0; }
+     double ScanningProgress() const { return fScanningWallet ? (double) m_scanning_progress : 0; }
```

re-utACK 469117d55c3601c75883d8ca7ff5e76f9f60da60

theStack commented on src/wallet/wallet.cpp:
```diff
@@ -1572,7 +1572,7 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
  * @return Earliest timestamp that could be successfully scanned from. Timestamp
  * returned will be higher than startTime if relevant blocks could not be read.
```
Now that this method doesn't return a timestamp (`int64_t`) anymore, this Doxygen comment needs to be updated.

Empact replied:
Updated, thanks!

Empact committed:
doc: Update CWallet::RescanFromTime return type documentation

And move to the header.

kallewoof commented on src/wallet/rpcdump.cpp:
```diff
@@ -1413,17 +1413,21 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
+            if (!pwallet->chain().findBlock(result.last_failed_block, nullptr /* block */, nullptr /* time */, &scannedTime)) {
+                throw std::logic_error("ScanForWalletTransactions returned invalid block hash");
+            }
+            scannedTime += TIMESTAMP_WINDOW + 1;
```
This differs from the original code, where `scannedTime` was set to the last block time + `TIMESTAMP_WINDOW + 1`. Won't this result in missed scans below, as `scannedTime` will be low?

Empact replied:
I don't see the difference in behavior - see the original location for comparison: https://github.com/bitcoin/bitcoin/pull/14942/commits/f059899fe3ca88204d9dda4f83c4f8c4548743ea#diff-b2bb174788c7409b671c46ccc86034bdL1607-L1611

Could you call out what you're referring to any more explicitly, in terms of what has changed?

kallewoof replied:
No, sorry, you're right. I missed the passing of `scannedTIme` on L#1427. It looks identical

kallewoof reviewed (commented):
Looks like an improvement. Left comments.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -169,6 +169,13 @@ std::shared_ptr<CWallet> LoadWallet(interfaces::Chain& chain, const std::string&
     return LoadWallet(chain, WalletLocation(name), error, warnings);
 }
 
+bool CWallet::AbortRescan()
```
Looks like you could have a race in this method?

Empact replied:
It's a bit convoluted, but I think it's handled gracefully: if `fScanningWallet` is set to `false` as `fAbortRescan` is being set to `true` (i.e. at the end of a scan), there is no issue as `ScanForWalletTransactions` sets `fAbortRescan` to `false` at its start.

BTW this behavior is unchanged - here's the prior implementation: https://github.com/bitcoin/bitcoin/pull/14942/files#diff-522490d83dce5375d423b23886e4125eL217-R219

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -214,9 +216,7 @@ UniValue abortrescan(const JSONRPCRequest& request)
-    if (!pwallet->IsScanning() || pwallet->IsAbortingRescan()) return false;
-    pwallet->AbortRescan();
-    return true;
+    return pwallet->AbortRescan();
```
Why does it matter the return value? Isn't even documented.

laanwj commented:
Any plans for this? It's needed rebase for more than a year, no other activity.

MarcoFalke commented:
Closing for now due to inactivity for more than a year. Can be reopened anytime. Just leave a comment here or in IRC.
